common4j/src/main/com/microsoft/identity/common/java/telemetry/Telemetry.java

```diff
@@ -0,0 +1,334 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.java.telemetry;
+
+import com.microsoft.identity.common.java.WarningType;
+import com.microsoft.identity.common.java.util.StringUtil;
+import com.microsoft.identity.common.java.telemetry.adapter.TelemetryAggregationAdapter;
+import com.microsoft.identity.common.java.telemetry.adapter.TelemetryDefaultAdapter;
+import com.microsoft.identity.common.java.telemetry.events.BaseEvent;
+import com.microsoft.identity.common.java.telemetry.observers.ITelemetryAggregatedObserver;
+import com.microsoft.identity.common.java.telemetry.observers.ITelemetryDefaultObserver;
+import com.microsoft.identity.common.java.telemetry.observers.ITelemetryObserver;
+import com.microsoft.identity.common.java.telemetry.rules.TelemetryPiiOiiRules;
+import com.microsoft.identity.common.java.logging.Logger;
+import com.microsoft.identity.common.java.logging.DiagnosticContext;
+
+import lombok.NonNull;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import static com.microsoft.identity.common.java.logging.DiagnosticContext.CORRELATION_ID;
+import static com.microsoft.identity.common.java.telemetry.TelemetryEventStrings.Key;
+
+/**
+ * A singleton class for logging Telemetry.
+ * Must be instantiated via {@link Telemetry.Builder} before use.
+ */
+public class Telemetry {
```

Comment:
General feedback on this class is that it seems like its method should be synchronized esp if we're creating, clearing, and initializing the collections inside it